README.md
IP 17mon java version, 比官方的速度快很多，支持监视文件改动自动加载

IP库请从 ipip.net 下载

添加依赖：
```
<dependency>
  <groupId>com.qiniu</groupId>
  <artifactId>ip17mon</artifactId>
  <version>0.2.1</version>
</dependency>
```

使用方法：

首选构建一个 `Locator` ，支持从不同的地方加载IP库，比如远程URL，或者本地的文件。

```
//从URL加载IP库
String ipBaseUrl = "http://xxx.example.com/ip.dat";
Locator locator = Locator.loadFromNet(ipBaseUrl);

//从本地加载IP库
String ipBasePath = "/tmp/ip.dat";
Locator locator = Locator.loadFromLocal(ipBasePath);
```
然后使用 `find` 方法查询：

```
LocationInfo ipInfo = locator.find("180.163.159.7");
System.out.println(ipInfo.toString());
```


CHANGELOG.md
# CHANGELOG

## 0.2.2 (2018-07-06)
### 增加
* 返回数组功能

## 0.2.1 (2018-01-02)
### 修正
* 去除重复代码

## 0.2.0 (2018-01-02)
### 增加
* 支持datx 格式

## 0.1.2 (2017-07-12)
### 增加
* 网络加载及流式加载ip库

## 0.0.2 (2017-01-016)
### 增加
* 自动更新加载
* 优化部分代码

## 0.0.1 (2017-01-016)
### 增加
* 高效IP查询实现



src/test/java/qiniu/ip17mon/LocationTest.java
package qiniu.ip17mon;

import org.junit.Assert;
import org.junit.Test;

import java.io.IOException;

/**
 * Created by long on 2017/1/16.
 */
public class LocationTest {
    @Test
    public void testLittleEndian() {
        byte[] x = new byte[]{1, 2, 3, 4};
        Assert.assertEquals(0x04030201, Locator.littleEndian(x, 0));
        byte[] x2 = new byte[]{0, 1, 2, 3, 4};
        Assert.assertEquals(0x04030201, Locator.littleEndian(x2, 1));
    }

    @Test
    public void testBigEndian() {
        byte[] x = new byte[]{1, 2, 3, 4};
        Assert.assertEquals(0x01020304, Locator.bigEndian(x, 0));
        byte[] x2 = new byte[]{0, 1, 2, 3, 4};
        Assert.assertEquals(0x01020304, Locator.bigEndian(x2, 1));
    }

    @Test
    public void testParseOctet() {
        Assert.assertEquals(1, Locator.parseOctet("1"));
        try {
            Locator.parseOctet("-1");
            Assert.fail();
        } catch (Exception e) {
            Assert.assertTrue(e instanceof NumberFormatException);
        }

        try {
            Locator.parseOctet("256");
            Assert.fail();
        } catch (Exception e) {
            Assert.assertTrue(e instanceof NumberFormatException);
        }
    }

    @Test
    public void testTextToNumericFormatV4() {
        byte[] b = Locator.textToNumericFormatV4("1.2.3.4");
        Assert.assertArrayEquals(new byte[]{1, 2, 3, 4}, b);

        try {
            Locator.textToNumericFormatV4("01.2.3.4");
        } catch (Exception e) {
            Assert.assertTrue(e instanceof NumberFormatException);
        }

        try {
            Locator.textToNumericFormatV4("2.3.4");
        } catch (Exception e) {
            Assert.assertTrue(e instanceof NumberFormatException);
        }

    }

    @Test
    public void testBuildLocationInfo() {
        String s = "中国\t上海\t上海\t浦东\t电信";
        LocationInfo l = Locator.buildInfo(s.getBytes(), 0, s.getBytes().length);
        Assert.assertEquals(new LocationInfo("中国", "上海", "上海", "电信"), l);
        Assert.assertArrayEquals(new String[]{"中国", "上海", "上海", "电信"}, l.toStringArray());
    }

    private void checkFreeLocator(Locator locator) {
        LocationInfo info = locator.find("8.8.8.8");
        Assert.assertEquals(new LocationInfo("GOOGLE", "GOOGLE", "", ""), info);
        info = locator.find("183.131.7.18");
        Assert.assertEquals(new LocationInfo("中国", "浙江", "杭州", ""), info);
        info = locator.find(new byte[]{(byte) 183, (byte) 131, 7, 18});
        Assert.assertEquals(new LocationInfo("中国", "浙江", "杭州", ""), info);
        info = locator.find(Locator.bigEndian(new byte[]{(byte) 183, (byte) 131, 7, 18}, 0));
        Assert.assertEquals(new LocationInfo("中国", "浙江", "杭州", ""), info);
    }

    private void checkFullLocator(Locator locator) {
        LocationInfo info = locator.find("8.8.8.8");
        Assert.assertEquals(new LocationInfo("GOOGLE.COM", "GOOGLE.COM", "", "level3.com"), info);
        info = locator.find("183.131.7.18");
        Assert.assertEquals(new LocationInfo("中国", "浙江", "杭州", "电信"), info);
        info = locator.find(new byte[]{(byte) 183, (byte) 131, 7, 18});
        Assert.assertEquals(new LocationInfo("中国", "浙江", "杭州", "电信"), info);
        info = locator.find(Locator.bigEndian(new byte[]{(byte) 183, (byte) 131, 7, 18}, 0));
        Assert.assertEquals(new LocationInfo("中国", "浙江", "杭州", "电信"), info);
    }

    @Test
    public void testNetLoadFree() throws IOException {
        Locator l = Locator.loadFromNet("http://7j1xnu.com1.z0.glb.clouddn.com/17monipdb.dat");
        checkFreeLocator(l);
    }

    @Test
    public void testLocalLoadFree() throws IOException {
        Locator l = Locator.loadFromLocal("17monipdb.dat");
        checkFreeLocator(l);
    }

//    @Test
//    public void testLocalLoadFull() throws IOException {
//        Locator l = Locator.loadFromLocal("mydata4vipday2.datx");
//        checkFullLocator(l);
//    }


    public void bench() throws IOException {
        Locator l = Locator.loadFromLocal("17monipdb.dat");
        long t1 = System.currentTimeMillis();
        l.checkDb();
        long t2 = System.currentTimeMillis();
        System.out.println("total time " + (t2 - t1));
        System.out.println("Ops " + (0xffffffffL * 1000 / (t2 - t1)));

    }
}


src/test/java/qiniu/ip17mon/AnotherLocator.java
package qiniu.ip17mon;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.Charset;
import java.util.InputMismatchException;

/**
 * Created by long on 2017/1/16.
 */
public final class AnotherLocator {
    private int offset;
    private int[] index = new int[256];
    private ByteBuffer dataBuffer;
    private ByteBuffer indexBuffer;

    private AnotherLocator(ByteBuffer buffer) {
        dataBuffer = buffer;

        dataBuffer.position(0);
        int indexLength = dataBuffer.getInt();
        byte[] indexBytes = new byte[indexLength];
        dataBuffer.get(indexBytes, 0, indexLength - 4);
        indexBuffer = ByteBuffer.wrap(indexBytes);
        indexBuffer.order(ByteOrder.LITTLE_ENDIAN);
        offset = indexLength;

        int loop = 0;
        while (loop++ < 256) {
            index[loop - 1] = indexBuffer.getInt();
        }
        indexBuffer.order(ByteOrder.BIG_ENDIAN);
    }

    private static long bytesToLong(byte a, byte b, byte c, byte d) {
        return int2long((((a & 0xff) << 24) | ((b & 0xff) << 16) | ((c & 0xff) << 8) | (d & 0xff)));
    }

    private static int str2Ip(String ip) {
        String[] ss = ip.split("\\.");
        int a, b, c, d;
        a = Integer.parseInt(ss[0]);
        b = Integer.parseInt(ss[1]);
        c = Integer.parseInt(ss[2]);
        d = Integer.parseInt(ss[3]);
        return (a << 24) | (b << 16) | (c << 8) | d;
    }

    private static long ip2long(String ip) {
        return int2long(str2Ip(ip));
    }

    private static long int2long(int i) {
        long l = i & 0x7fffffffL;
        if (i < 0) {
            l |= 0x080000000L;
        }
        return l;
    }

    public static AnotherLocator loadFromNet(String netPath) throws IOException {
        URL url = new URL(netPath);
        HttpURLConnection httpConn = (HttpURLConnection) url.openConnection();
        httpConn.setConnectTimeout(3000);
        httpConn.setReadTimeout(30 * 1000);
        int responseCode = httpConn.getResponseCode();
        if (responseCode != HttpURLConnection.HTTP_OK) {
            return null;
        }

        int length = httpConn.getContentLength();
        if (length <= 0 || length > 10 * 1024 * 1024) {
            throw new InputMismatchException("invalid ip data");
        }
        InputStream is = httpConn.getInputStream();
        byte[] data = new byte[length];
        int downloaded = 0;
        while (downloaded < length) {
            int read = is.read(data, downloaded, length - downloaded);
            if (read < 0) {
                is.close();
                throw new IOException("read error");
            }
            downloaded += read;
        }

        is.close();

        ByteBuffer bb = ByteBuffer.wrap(data);
        return new AnotherLocator(bb);
    }

    public static AnotherLocator loadFromLocal(String filePath) throws IOException {
        File f = new File(filePath);
        FileInputStream fi = new FileInputStream(f);
        ByteBuffer bb = ByteBuffer.allocate((int) f.length());

        try {
            fi.read(bb.array());
        } catch (IOException e) {
            e.printStackTrace();
            try {
                fi.close();
            } catch (IOException e1) {
                e1.printStackTrace();
            }
            throw e;
        }
        fi.close();

        return new AnotherLocator(bb);
    }

    public String[] find(String ip) {
        int ip_prefix_value = new Integer(ip.substring(0, ip.indexOf(".")));
        long ip2long_value = ip2long(ip);
        int start = index[ip_prefix_value];
        int max_comp_len = offset - 1028;
        long index_offset = -1;
        int index_length = -1;
        byte b = 0;
        for (start = start * 8 + 1024; start < max_comp_len; start += 8) {
            if (int2long(indexBuffer.getInt(start)) >= ip2long_value) {
                index_offset = bytesToLong(b, indexBuffer.get(start + 6), indexBuffer.get(start + 5),
                        indexBuffer.get(start + 4));
                index_length = 0xFF & indexBuffer.get(start + 7);
                break;
            }
        }

        byte[] areaBytes;

        dataBuffer.position(offset + (int) index_offset - 1024);
        areaBytes = new byte[index_length];
        dataBuffer.get(areaBytes, 0, index_length);

        return new String(areaBytes, Charset.forName("UTF-8")).split("\t", -1);
    }
}


src/main/java/qiniu/ip17mon/LocationInfo.java
package qiniu.ip17mon;

/**
 * Created by long on 2017/1/16.
 */
public final class LocationInfo {
    public final String country;
    public final String state;
    public final String city;
    public final String isp;


    public LocationInfo(String country, String state, String city, String isp) {
        this.country = country;
        this.state = state;
        this.isp = isp;
        this.city = city;
    }

    public boolean equals(Object o) {
        if (o == null || !(o instanceof LocationInfo)) {
            return false;
        }
        LocationInfo l = (LocationInfo) o;
        return country.equals(l.country) && state.equals(l.state) && city.equals(l.city) && isp.equals(l.isp);
    }

    public int hashCode() {
        return country.hashCode() * 31 * 31 * 31 + state.hashCode() * 31 * 31 + city.hashCode() * 31 + isp.hashCode();
    }

    public String toString() {
        return country + " " + state + " " + city + " " + isp;
    }

    public String[] toStringArray() {
        return new String[]{country, state, city, isp};
    }

}


src/main/java/qiniu/ip17mon/AutoReloadLocator.java
package qiniu.ip17mon;

import java.io.IOException;
import java.util.Observable;
import java.util.Observer;

/**
 * Created by long on 2017/1/17.
 */
public final class AutoReloadLocator implements ILocator, Observer {

    private final SimpleFileWatchService watchService;
    private final String filePath;
    private Locator locator;

    public AutoReloadLocator(String filePath, int intervalSeconds) throws IOException {
        this.filePath = filePath;
        locator = Locator.loadFromLocal(filePath);
        watchService = new SimpleFileWatchService(filePath, intervalSeconds);
        watchService.addObserver(this);
        watchService.excute();
    }

    // only for test
    public static void main(String[] args) {
        String filePath = "17monipdb.dat";
        if (args != null && args.length > 0) {
            filePath = args[0];
        }
        AutoReloadLocator l;
        try {
            l = new AutoReloadLocator(filePath, 10);
            System.out.println(l.find("8.8.8.8"));
        } catch (IOException e) {
            e.printStackTrace();
        }
        try {
            Thread.sleep(60 * 60 * 1000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    @Override
    public LocationInfo find(String ip) {
        return locator.find(ip);
    }

    @Override
    public LocationInfo find(byte[] ipBin) {
        return locator.find(ipBin);
    }

    @Override
    public LocationInfo find(int address) {
        return locator.find(address);
    }

    @Override
    public void update(Observable o, Object arg) {
        try {
            locator = Locator.loadFromLocal(filePath);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void shutdown() {
        watchService.deleteObserver(this);
        watchService.shutdown();
    }
}


src/main/java/qiniu/ip17mon/SimpleFileWatchService.java
package qiniu.ip17mon;

import java.io.File;
import java.util.Observable;

/**
 * Created by long on 2017/1/17.
 */
final class SimpleFileWatchService extends Observable implements Runnable {
    private File file;
    private Thread thread;
    private long lastModified = 0;
    private int intervalSeconds;
    private volatile boolean flag;

    SimpleFileWatchService(String filePath, int intervalSeconds) {
        this.file = new File(filePath);
        this.intervalSeconds = intervalSeconds;
        this.lastModified = file.lastModified();
        this.thread = new Thread(this);
        this.flag = true;
    }

    private void monitorFile() {
        //delay monitor
        try {
            Thread.sleep(5 * 60 * 1000);
        } catch (InterruptedException e) {
//            e.printStackTrace();
        }
        while (flag) {
            long l = file.lastModified();
            if (l != lastModified) {
                lastModified = l;
                setChanged();
                notifyObservers();
            }
            try {
                Thread.sleep(intervalSeconds * 1000);
            } catch (InterruptedException e) {
//                e.printStackTrace();
            }
        }
    }

    void excute() {
        thread.start();
    }

    @Override
    public void run() {
        monitorFile();
    }

    public void shutdown() {
        flag = false;
    }
}


src/main/java/qiniu/ip17mon/Locator.java
package qiniu.ip17mon;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.Charset;
import java.util.InputMismatchException;

public final class Locator implements ILocator {
    public static final String VERSION = "0.2.2";
    private static final Charset Utf8 = Charset.forName("UTF-8");
    private final byte[] ipData;
    private final int textOffset;
    private final int[] index;
    private final int[] indexData;
    private final int[] textStartIndex;
    private final short[] textLengthIndex;

    private Locator(byte[] data) {
        this(data, (data[4] == 0 && data[5] == 0 && data[6] == 0 && data[7] == 0) && data[8] == 0 && data[9] == 0);
    }

    private Locator(byte[] data, boolean datx) {
        this.ipData = data;
        int offset = bigEndian(data, 0);
        this.index = new int[256];
        if (datx) {

            textOffset = offset - 256 * 256 * 4;

            for (int i = 0; i < 256; i++) {
                index[i] = littleEndian(data, 4 + i * 4 * 256);
            }

            int nidx = (textOffset - 4 - 256 * 256 * 4) / 9;
            indexData = new int[nidx];
            textStartIndex = new int[nidx];
            textLengthIndex = new short[nidx];

            for (int i = 0, off = 0; i < nidx; i++) {
                off = 4 + 256 * 256 * 4 + i * 9;
                indexData[i] = bigEndian(ipData, off);
                textStartIndex[i] = ((int) ipData[off + 6] & 0xff) << 16 | ((int) ipData[off + 5] & 0xff) << 8
                        | ((int) ipData[off + 4] & 0xff);
                textLengthIndex[i] = (short) (((int) ipData[off + 7] & 0xff) << 8 | ((int) ipData[off + 8] & 0xff));
            }
        } else {

            for (int i = 0; i < 256; i++) {
                index[i] = littleEndian(data, 4 + i * 4);
            }
            textOffset = offset - 1024;

            int nIdx = (textOffset - 4 - 1024) / 8;
            indexData = new int[nIdx];
            textStartIndex = new int[nIdx];
            textLengthIndex = new short[nIdx];

            for (int i = 0, off = 0; i < nIdx; i++) {
                off = 4 + 1024 + i * 8;
                indexData[i] = bigEndian(ipData, off);
                textStartIndex[i] = ((int) ipData[off + 6] & 0xff) << 16 | ((int) ipData[off + 5] & 0xff) << 8
                        | ((int) ipData[off + 4] & 0xff);
                textLengthIndex[i] = ipData[off + 7];
            }
        }
    }

    static int bigEndian(byte[] data, int offset) {
        int a = (((int) data[offset]) & 0xff);
        int b = (((int) data[offset + 1]) & 0xff);
        int c = (((int) data[offset + 2]) & 0xff);
        int d = (((int) data[offset + 3]) & 0xff);
        return (a << 24) | (b << 16) | (c << 8) | d;
    }

    static int littleEndian(byte[] data, int offset) {
        int a = (((int) data[offset]) & 0xff);
        int b = (((int) data[offset + 1]) & 0xff);
        int c = (((int) data[offset + 2]) & 0xff);
        int d = (((int) data[offset + 3]) & 0xff);
        return (d << 24) | (c << 16) | (b << 8) | a;
    }

    static byte parseOctet(String ipPart) {
        // Note: we already verified that this string contains only hex digits.
        int octet = Integer.parseInt(ipPart);
        // Disallow leading zeroes, because no clear standard exists on
        // whether these should be interpreted as decimal or octal.
        if (octet < 0 || octet > 255 || (ipPart.startsWith("0") && ipPart.length() > 1)) {
            throw new NumberFormatException("invalid ip part");
        }
        return (byte) octet;
    }

    static byte[] textToNumericFormatV4(String str) {
        String[] s = str.split("\\.");
        if (s.length != 4) {
            throw new NumberFormatException("the ip is not v4");
        }
        byte[] b = new byte[4];
        b[0] = parseOctet(s[0]);
        b[1] = parseOctet(s[1]);
        b[2] = parseOctet(s[2]);
        b[3] = parseOctet(s[3]);
        return b;
    }

    static LocationInfo buildInfo(byte[] bytes, int offset, int len) {
        String str = new String(bytes, offset, len, Utf8);
        String[] ss = str.split("\t", -1);
        if (ss.length == 4) {
            return new LocationInfo(ss[0], ss[1], ss[2], "");
        } else if (ss.length >= 5) {
            return new LocationInfo(ss[0], ss[1], ss[2], ss[4]);
        } else if (ss.length == 3) {
            return new LocationInfo(ss[0], ss[1], ss[2], "");
        } else if (ss.length == 2) {
            return new LocationInfo(ss[0], ss[1], "", "");
        }
        return null;
    }

    public static Locator loadFromNet(String netPath) throws IOException {
        URL url = new URL(netPath);
        HttpURLConnection httpConn = (HttpURLConnection) url.openConnection();
        httpConn.setConnectTimeout(3000);
        httpConn.setReadTimeout(30 * 1000);
        int responseCode = httpConn.getResponseCode();
        if (responseCode != HttpURLConnection.HTTP_OK) {
            return null;
        }

        int length = httpConn.getContentLength();
        if (length <= 0 || length > 64 * 1024 * 1024) {
            throw new InputMismatchException("invalid ip data");
        }
        InputStream is = httpConn.getInputStream();
        byte[] data = new byte[length];
        int downloaded = 0;
        int read = 0;
        while (downloaded < length) {
            try {
                read = is.read(data, downloaded, length - downloaded);
            } catch (IOException e) {
                is.close();
                throw new IOException("read error");
            }
            if (read < 0) {
                is.close();
                throw new IOException("read error");
            }
            downloaded += read;
        }

        is.close();

        String path = url.getPath();
        if (path.toLowerCase().endsWith("datx")) {
            return loadBinary(data, true);
        } else if (path.toLowerCase().endsWith("dat")) {
            return loadBinary(data, false);
        } else {
            return loadBinaryUnkown(data);
        }
    }

    public static Locator loadFromLocal(String filePath) throws IOException {
        File f = new File(filePath);
        FileInputStream fi = new FileInputStream(f);
        byte[] b = new byte[(int) f.length()];
        try {
            fi.read(b);
        } catch (IOException e) {
            e.printStackTrace();
            try {
                fi.close();
            } catch (IOException e1) {
                e1.printStackTrace();
            }
            throw e;
        }
        fi.close();

        if (filePath.toLowerCase().endsWith("datx")) {
            return loadBinary(b, true);
        } else if (filePath.toLowerCase().endsWith("dat")) {
            return loadBinary(b, false);
        } else {
            return loadBinaryUnkown(b);
        }
    }

    public static Locator loadFromStream(InputStream in, boolean x) throws Exception {
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        byte[] buffer = new byte[16 * 1024];
        int n;

        while ((n = in.read(buffer)) != -1) {
            byteArrayOutputStream.write(buffer, 0, n);
        }

        return loadBinary(byteArrayOutputStream.toByteArray(), x);
    }

    public static Locator loadFromStreamUnkown(InputStream in) throws Exception {
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        byte[] buffer = new byte[16 * 1024];
        int n;

        while ((n = in.read(buffer)) != -1) {
            byteArrayOutputStream.write(buffer, 0, n);
        }

        return loadBinaryUnkown(byteArrayOutputStream.toByteArray());
    }

    public static Locator loadFromStreamOld(InputStream in) throws Exception {
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        byte[] buffer = new byte[16 * 1024];
        int n;

        while ((n = in.read(buffer)) != -1) {
            byteArrayOutputStream.write(buffer, 0, n);
        }

        return loadBinaryOld(byteArrayOutputStream.toByteArray());
    }

    public static Locator loadBinary(byte[] ipdb, boolean x) {
        return new Locator(ipdb, x);
    }

    public static Locator loadBinaryUnkown(byte[] ipdb) {
        return new Locator(ipdb);
    }

    public static Locator loadBinaryOld(byte[] ipdb) {
        return loadBinary(ipdb, false);
    }

    public static void main(String[] args) {
        if (args == null || args.length < 2) {
            System.out.println("locator ipfile ip");
            return;
        }
        try {
            Locator l = loadFromLocal(args[0]);
            System.out.println(l.find(args[1]));
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    private int findIndexOffset(long ip, int start, int end) {
        int mid = 0;
        while (start < end) {
            mid = (start + end) / 2;
            long l = 0xffffffffL & ((long) indexData[mid]);
            if (ip > l) {
                start = mid + 1;
            } else {
                end = mid;
            }
        }
        long l = ((long) indexData[end]) & 0xffffffffL;
        if (l >= ip) {
            return end;
        }
        return start;
    }

    public LocationInfo find(String ip) {
        byte[] b;
        try {
            b = textToNumericFormatV4(ip);
        } catch (Exception e) {
            return null;
        }
        return find(b);
    }

    public LocationInfo find(byte[] ipBin) {
        int end = indexData.length - 1;
        int a = 0xff & ((int) ipBin[0]);
        if (a != 0xff) {
            end = index[a + 1];
        }
        long ip = (long) bigEndian(ipBin, 0) & 0xffffffffL;
        int idx = findIndexOffset(ip, index[a], end);
        int off = textStartIndex[idx];
        return buildInfo(ipData, textOffset + off, 0xffff & (int) textLengthIndex[idx]);
    }

    public LocationInfo find(int address) {
        byte[] addr = new byte[4];

        addr[0] = (byte) ((address >> 24) & 0xff);
        addr[1] = (byte) ((address >> 16) & 0xff);
        addr[2] = (byte) ((address >> 8) & 0xff);
        addr[3] = (byte) (address & 0xff);

        return find(addr);
    }

    public void checkDb() throws IOException {
        byte[] addr = new byte[4];
        try {
            for (long x = 0; x < 0xffffffffL; x++) {
                addr[0] = (byte) ((x >> 24) & 0xff);
                addr[1] = (byte) ((x >> 16) & 0xff);
                addr[2] = (byte) ((x >> 8) & 0xff);
                addr[3] = (byte) (x & 0xff);
                find(addr);
            }
        } catch (Exception e) {
            throw new IOException(e.getMessage());
        }
    }
}
